### ISSUE TITLE
[Bug]: unexpected duplicate outputs triggered by quota exceeded exception in finish_bundle() 

### ISSUE THREAD

**Issue by fathom-zhou** [apache/beam#27156]
### What happened?

Beam version:  2.46.0
Language: Python
Runner: Dataflow

I am encountering an issue with my BEAM pipeline, which involves performing transformations on a set of documents and saving them into TFRecord. Within the pipeline, I have incorporated a DoFn that logs the transformation info using Stackdriver logging. However, when a quota exceeded exception is raised in the finish_bundle phase of DoFn, the pipeline doesn't fail and some documents end up being saved multiple times, leading to unexpected duplication. This behaviour may be a bug in BEAM or Dataflow.

My current solution is add retry in the finish_bundle whenever quota issue is encountered and this successfully prevents the duplicated docs. 

Due to the proprietary restrictions, I am not able to share the source code here. Please let me know if you need more details about this issue. 

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [X] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [X] Component: Google Cloud Dataflow Runner

**Comment by tvalentyn** [comment]
it sounds like your `process` and/or `finish_bundle` methods may be not idempotent. Idempotency is an important consideration when writing IO code. Are you using TFRecordIO by chance?  If not: where is saving the document part is happening?
 https://github.com/apache/beam/blob/ca674aad086e4d1a40e8c01598073030a22484fa/sdks/python/apache_beam/io/tfrecordio.py

**Comment by tvalentyn** [comment]
Friendly reminder re: the above question. Thanks!

**Comment by kvudata** [comment]
Yes, we are using TFRecordIO (specifically `beam.io.WriteToTFRecord`).

> it sounds like your process and/or finish_bundle methods may be not idempotent. Idempotency is an important consideration when writing IO code.

Note that we're only logging to Stackdriver aka Google Cloud Logging in our DoFn (we gather logs in `process()` and then log the batch in `finish_bundle()`), and we're ok with duplicate logs being generated. We're observing that Dataflow seems to perform some kind of retry if our `finish_bundle()` fails, and the later `WriteToTFRecord` (which doesn't depend on this DoFn in our pipeline) ends up writing duplicates.

Another interesting observation is that the metrics for the number of items processed / written in the Dataflow UI is the number of elements we would expect if duplicates were not being written - the duplicates are only apparent from inspecting the output tfrecord(s).

**Comment by tvalentyn** [comment]
The behavior you are describing is surprising. A failure in `finish_bundle` results in bundle retry, but that should not result in elements written twice in the IO step. I do not reproduce the behavior described here. I tried running the following pipeline on Dataflow:

```
import argparse
import logging

import apache_beam as beam
from apache_beam.options.pipeline_options import PipelineOptions
from apache_beam.io.gcp.pubsub import ReadFromPubSub
from apache_beam.testing.synthetic_pipeline import SyntheticSource

import logging
_LOGGER = logging.getLogger(__name__)
tf_out_path = 'gs://clouddfe-valentyn/finish_repro_10m'

def tf_record_finish_bundle(p):
  class SometimesFailFn(beam.DoFn):
    def start_bundle(self):
      global bundle_counter
      _LOGGER.warning("Start processing bundle")

    def finish_bundle(self):
      from random import random
      if random() < 0.25:
        raise RuntimeError(f"Intentionally failing finish_bundle.")

      _LOGGER.warning("Finished processing bundle")

    def process(self, element):
      yield element[1]

  (
    p
    | beam.io.Read(
              SyntheticSource({
                  "numRecords": 10000000, "keySizeBytes": 1, "valueSizeBytes": 1
              }))

   | beam.Reshuffle()
   | beam.ParDo(SometimesFailFn())
   | beam.io.tfrecordio.WriteToTFRecord(tf_out_path)
  )


def verify_tfrecord_output(p):
  (p
   | beam.io.tfrecordio.ReadFromTFRecord(tf_out_path+"*")
   | beam.combiners.Count.Globally()
   | "LogCount" >> beam.LogElements(prefix="TotalCount ", level=logging.WARNING)
  )

def run(argv=None):
  parser = argparse.ArgumentParser()
  known_args, pipeline_args = parser.parse_known_args(argv)
  print(known_args)
  print(pipeline_args)

  with beam.Pipeline(argv=pipeline_args) as p:
    tf_record_finish_bundle(p)

  with beam.Pipeline(argv=pipeline_args) as p:
    verify_tfrecord_output(p)

```

I verified that there was at least 1 "Intentionally failing finish_bundle", but the total number of elements written in TFRecordIO was as expected.

**Comment by tvalentyn** [comment]
Would you be able to provide a repro that demonstrates that duplicate elements are written?
